pfcpiface/p4rt_translator.go

```diff
@@ -407,6 +503,71 @@ func (t *P4rtTranslator) ParseAccessIPFromReadInterfaceTableResponse(resp *p4.Re
 	return nil, ErrOperationFailed("parse Access IP from P4Runtime response")
 }
 
+func (t *P4rtTranslator) BuildApplicationsTableEntry(pdr pdr, internalAppID uint8) (*p4.TableEntry, error) {
+	applicationsBuilderLog := log.WithFields(log.Fields{
+		"pdr": pdr,
+	})
+	applicationsBuilderLog.Trace("Building P4rt table entry for applications table")
+
+	tableID := t.tableID(TableApplications)
+	entry := &p4.TableEntry{
+		TableId: tableID,
+		// priority for UP4 cannot be greater than 65535
+		Priority: int32(math.MaxUint8 - pdr.precedence),
+	}
+
+	var (
+		appIP, appIPMask uint32 = 0, 0
+		appPort          uint16 = 0
+	)
+
+	if pdr.srcIface == access {
+		appIP, appIPMask = pdr.appFilter.dstIP, pdr.appFilter.dstIPMask
+		appPort = pdr.appFilter.dstPort
+	} else if pdr.srcIface == core {
+		appIP, appIPMask = pdr.appFilter.srcIP, pdr.appFilter.srcIPMask
+		appPort = pdr.appFilter.srcPort
+	}
+
+	appProto, appProtoMask := pdr.appFilter.proto, pdr.appFilter.protoMask
+
+	appIPPrefixLen := 32 - bits.TrailingZeros32(appIPMask)
```

Comment:
Is there any guarantee that the mask is equivalent to a prefix lenght? Can there be other zeros besides the trailing ones?

Reply:
This is guaranteed by the format of SDF Filter that uses `IP/prefixlen`.